### PR TITLE
Fix charRange to allow picking the last item

### PR DESCRIPTION
**Pull request by PunKeel**
Before this change, choose would never pick the last character of the set. This is not the expected behavior: if the range is a-z, we expect a,b,...,y,z to be picked in a uniform way.

### CODE REVIEW

**Comment by PunKeel** [comment]
@lavalamp Could you please rerun the tests on the second commit for Go 1.13?

File: fuzz_test.go [hunk 1]
```diff
@@ -504,3 +505,37 @@ func TestFuzz_SkipPattern(t *testing.T) {
 		return 5, true
 	})
 }
+
+func Test_charRange_choose(t *testing.T) {
+	// We run 10,000 iterations of charRange.choose and verify 'a', 'z' (edge cases) and 'm'
```
Comment (lavalamp) [inline fuzz_test.go]: Suggestion: make a fake rand, e.g.:

```
type alwaysLast struct {
  *rand.Rand
}

func (a alwaysLast) Int63n(n int64) int64 {
  return n-1
}
```

(you can imagine various extensions to this to test other edge cases)

Then you can test the edge cases instead of trying a bunch of times.

File: fuzz.go [hunk 2]
```diff
@@ -478,7 +478,7 @@ type charRange struct {
 // choose returns a random unicode character from the given range, using the
 // given randomness source.
 func (r *charRange) choose(rand *rand.Rand) rune {
```
Comment (lavalamp) [inline fuzz.go]: I know it's unrelated, but can you change this to `r *rand.Rand`? Hiding the package name with the variable name is confusing and gave me a minor panic attack just now because this code should not be making calls to the built in rand package loose functions.

**Comment by lavalamp** [comment]
Thanks for the fix, I have a suggestion!

File: fuzz.go (hunk 2 above)
Comment (PunKeel) [reply fuzz.go]: done

File: fuzz_test.go (hunk 1 above)
Comment (PunKeel) [reply fuzz_test.go]: done

**Comment by PunKeel** [comment]
Thanks for the suggestions!

I could not find a way to pass a mocked *rand.Rand, so I've introduced a new interface `int63nPicker`, I hope this is acceptable.

**Comment by PunKeel** [comment]
Bonus: added a constant seed to all tests, to (hopefully) help stabilize them.

File: fuzz_test.go [hunk 3]
```diff
@@ -504,3 +532,37 @@ func TestFuzz_SkipPattern(t *testing.T) {
...
+}
+
+func (a alwaysLast) Int63n(n int64) int64 {
+	return n - 1
+}
+
+type alwaysFirst struct {
+}
+
+func (a alwaysFirst) Int63n(n int64) int64 {
```
Comment (lavalamp) [inline fuzz_test.go]: I don't super object to making the interface like you did, but go's embedding feature makes it pretty easy to produce a *rand.Rand with only a single method overridden like this:

```
type customInt63 struct {
  *rand.Rand

  mode int63mode
}

func (c customInt63) Int63n(n int64) int64 {
  switch c.mode {
  case modeFirst: return 0
  case modeLast: return n-1
  default: return c.Rand.Int63n(n)
  }
}
```

File: fuzz_test.go [hunk 4]
```diff
@@ -42,8 +43,10 @@ func TestFuzz_basic(t *testing.T) {
 	}{}
 
 	failed := map[string]int{}
+	f := New().
+		RandSource(rand.NewSource(10))
```
Comment (lavalamp) [inline fuzz_test.go]: I appreciate the desire to make the tests deterministic, but I'm torn-- I want them to pass with all seeds (and I think they don't right now), I'm worried that sticking to a seed that makes them pass might obscure a legitimate problem. Can you separate these changes into a separate PR so we can get your original fix in sooner? :)

Comment (PunKeel) [reply fuzz_test.go]: Sure can - I included it in this PR as the tests were "randomly" failing

File: fuzz_test.go (hunk 3 above)
Comment (PunKeel) [reply fuzz_test.go]: I couldn't get the `type x struct { *rand.Rand }` code to work, hence why I introduced the interface. I'll try harder

Comment (lavalamp) [reply fuzz_test.go]: Oh wait-- rand.Rand isn't an interface. You're right, that's not going to work.

Comment (lavalamp) [reply fuzz_test.go]: Yeah let's stick with the interface you made and just split the seeding of the tests into another PR and we can merge this.

Comment (PunKeel) [reply fuzz_test.go]: Switched to use a custom object with a "mode" instead of two objects.

File: fuzz_test.go (hunk 4 above)
Comment (PunKeel) [reply fuzz_test.go]: removed

**Comment by lavalamp** [comment]
LGTM, rerunning the test
